app/assets/javascripts/discourse/app/components/topic-list/topic-list-item.gjs

```diff
@@ -78,6 +76,26 @@ export default class TopicListItem extends Component {
     return this.site.desktopView && this.args.focusLastVisitedTopic;
   }
 
+  @cached
+  get columns() {
+    const self = this;
+    const context = {
+      get category() {
+        return self.topicTrackingState.get("filterCategory");
+      },
+
+      get filter() {
+        return self.topicTrackingState.get("filter");
+      },
+    };
+
+    return applyValueTransformer(
+      "topic-list-columns",
+      createColumns(),
+      context
+    );
```

Comment:
Seems like we're re-building the DAG from scratch for the header, and then for every single row. For perf, I think it would make more sense to generate it at the top TopicList level, and then pass the resolved DAG down into each row?